File: C++/Palindrome-Partitioning.cpp
class Solution {
public:
    void helper(vector<vector<string> >& res, vector<vector<int> >& sign, vector<string>& cur, int pos, string &s) {
        if (pos == s.size()) {
            res.push_back(cur);
            return;
        }
        for (int i = pos; i < s.size(); i ++) {
            if (sign[pos][i] == 1) {
                cur.push_back(s.substr(pos, i - pos + 1));
                helper(res, sign, cur, i + 1, s);
                cur.pop_back();
            }
        }
    }
    vector<vector<string> > partition(string s) {
        vector<vector<string> > res;
        if (s.empty()) return res;
        int len = s.size();
        vector<vector<int> > sign(len, vector<int>(len, 0));
        for (int i = 1; i <= len; i ++) {
            for (int j = 0; j + i - 1 < len; j ++) {
                if (i == 1) sign[j][j] = 1;
                else {
                    int k = j + i - 1;
                    if ((j + 1 >= k - 1 || sign[j + 1][k - 1] == 1) && s[j] == s[k]) {
                        sign[j][k] = 1;
                    }
                }
            }
        }
        vector<string> cur;
        helper(res, sign, cur, 0, s);
        return res;
    }
};

File: Python/Palindrome-Partitioning.py
import copy

class Solution:
    # @param s, a string
    # @return a list of lists of string
    def isPalindrome(self, s, left, right):
        if left == right:
            return True
        mid = (left + right) / 2
        for i in range(left, mid + 1):
            if s[i] != s[right - i + left]:
                return False
        return True
    def generate(self, s, pos, end, cnt, res, seq):
        if end == len(s) and pos == len(s):
            res.append(copy.copy(cnt))
            return
        if end >= len(s):
            return
        if seq[pos][end] == 1:
            cnt.append(s[pos:end+1])
            self.generate(s, end+1, end+1, cnt, res, seq)
            cnt.pop()
        self.generate(s, pos, end+1, cnt, res, seq)
        return
    def partition(self, s):
        length = len(s)
        seq = [[0 for col in range(length)] for row in range(length)]
        for i in range(length):
            for j in range(i + 1):
                if self.isPalindrome(s, j, i):
                    seq[j][i] = 1
        cnt = []
        res = []
        self.generate(s, 0, 0, cnt, res, seq)
        return res